tests/unittests/Foundation/NSDecimalNumberTests.mm

```diff
@@ -0,0 +1,330 @@
+    EXPECT_EQ(0, result._length);
+    EXPECT_EQ(0, result._isNegative);
+    //
+    left = [[[[NSDecimalNumber alloc] initWithUnsignedLongLong:9223372534876] autorelease] decimalValue];
+    right = [[[[NSDecimalNumber alloc] initWithDouble:0] autorelease] decimalValue];
+
+    NSDecimalSubtract(&result, &left, &right, 0);
+    EXPECT_EQ(NSOrderedSame, NSDecimalCompare(&result, &left));
+
+    //
+    unsigned short mantissa[NSDecimalMaxSize] = { 2048, 25428, 61485, 43339, 53, 0, 0, 0 };
+    result = { 0 };
+    left = [[[[NSDecimalNumber alloc] initWithDouble:-10123456e-6] autorelease] decimalValue];
+    right = [[[[NSDecimalNumber alloc] initWithDouble:-1e3] autorelease] decimalValue];
+
+    NSDecimalSubtract(&result, &left, &right, 0);
+    EXPECT_ARRAY_EQUAL(mantissa, result._mantissa);
+    EXPECT_EQ(5, result._length);
+    EXPECT_EQ(0, result._isNegative);
+    EXPECT_EQ(-18, result._exponent);
+
+    //
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:10] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:0.5] autorelease]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:1000] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-989.5] autorelease]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:-10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:10] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-20.5] autorelease]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:-1e3] autorelease],
+                 [[[NSDecimalNumber alloc] initWithDouble:1010.5] autorelease]);
+
+    NSDecimalNumber* d = [[[NSDecimalNumber alloc] initWithDouble:1012345678912345e-120] autorelease];
+    testSubtract(d, d, [NSDecimalNumber zero]);
+
+    testSubtract([[[NSDecimalNumber alloc] initWithInteger:2] autorelease], [NSDecimalNumber one], [NSDecimalNumber one]);
+}
+
+TEST(NSDecimalNumber, addition) {
+    // Note: Subtraction tests addition, since NSDecimalSubtract gets converted to NSDecimalAdd
+    unsigned short mantissa[NSDecimalMaxSize] = { 5272, 10, 0, 0, 0, 0, 0, 0 };
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithLongLong:6560451.0] autorelease];
+    NSDecimal decimal1 = [dNumber decimalValue];
+
+    dNumber = [[[NSDecimalNumber alloc] initWithDouble:45869] autorelease];
+    double maxx = DBL_MAX;
+    NSDecimal decimal2 = [dNumber decimalValue];
+
+    NSDecimal result = { 0 };
+
+    NSDecimalAdd(&result, &decimal1, &decimal2, 0);
+
+    ASSERT_EQ(2, result._length);
+    ASSERT_EQ(1, result._exponent);
+    ASSERT_EQ(0, result._isNegative);
+
+    EXPECT_ARRAY_EQUAL(mantissa, result._mantissa);
```

Comment:
Is this really testing the implementation details?  will it pass on reference platform?  You should split the implementation details tests into windows only test.